pallets/funding/src/runtime_api.rs

```diff
@@ -0,0 +1,225 @@
+	}
+
+	#[api_version(1)]
+	pub trait ExtrinsicHelpers<T: Config> {
+		/// Get the current price of a contribution token (either current bucket in the auction, or WAP in contribution phase),
+		/// and calculate the amount of tokens that can be bought with the given amount USDT/USDC/DOT.
+		fn funding_asset_to_ct_amount(project_id: ProjectId, asset: AcceptedFundingAsset, asset_amount: BalanceOf<T>) -> BalanceOf<T>;
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn top_evaluators(project_id: ProjectId, amount: u32) -> Vec<EvaluationInfoOf<T>> {
+		Evaluations::<T>::iter_prefix_values((project_id,))
+			.sorted_by(|a, b| b.original_plmc_bond.cmp(&a.original_plmc_bond))
+			.take(amount as usize)
+			.collect_vec()
+	}
+
+	pub fn top_bidders(project_id: ProjectId, amount: u32) -> Vec<BidInfoOf<T>> {
+		Bids::<T>::iter_prefix_values((project_id,))
+			.sorted_by(|a, b| b.final_ct_amount.cmp(&a.final_ct_amount))
+			.take(amount as usize)
+			.collect_vec()
+	}
+
+	pub fn top_contributors(project_id: ProjectId, amount: u32) -> Vec<ContributionInfoOf<T>> {
+		Contributions::<T>::iter_prefix_values((project_id,))
+			.sorted_by(|a, b| b.ct_amount.cmp(&a.ct_amount))
+			.take(amount as usize)
+			.collect_vec()
+	}
+
+	pub fn top_projects_by_usd_raised(amount: u32) -> Vec<(ProjectId, ProjectMetadataOf<T>, ProjectDetailsOf<T>)> {
+		ProjectsDetails::<T>::iter()
+			.sorted_by(|a, b| b.1.funding_amount_reached_usd.cmp(&a.1.funding_amount_reached_usd))
+			.take(amount as usize)
+			.map(|(project_id, project_details)| {
+				let project_metadata = ProjectsMetadata::<T>::get(project_id).unwrap();
+				(project_id, project_metadata, project_details)
+			})
+			.collect_vec()
+	}
+
+	pub fn top_projects_by_usd_target_percent_reached(
+		amount: u32,
+	) -> Vec<(ProjectId, ProjectMetadataOf<T>, ProjectDetailsOf<T>)> {
+		ProjectsDetails::<T>::iter()
+			.map(|(project_id, project_details)| {
+				let funding_reached = project_details.funding_amount_reached_usd;
+				let funding_target = project_details.fundraising_target_usd;
+				let funding_ratio = FixedU128::from_rational(funding_reached.into(), funding_target.into());
+				(project_id, project_details, funding_ratio)
+			})
+			.sorted_by(|a, b| b.2.cmp(&a.2))
+			.take(amount as usize)
+			.map(|(project_id, project_details, _funding_ratio)| {
+				let project_metadata = ProjectsMetadata::<T>::get(project_id).unwrap();
+				(project_id, project_metadata, project_details)
+			})
+			.collect_vec()
+	}
+
+	pub fn contribution_tokens(account: AccountIdOf<T>) -> Vec<(ProjectId, BalanceOf<T>)> {
+		let asset_ids = <T as Config>::ContributionTokenCurrency::asset_ids();
+		asset_ids
+			.filter_map(|asset_id| {
+				let balance = <T as Config>::ContributionTokenCurrency::balance(asset_id, &account);
+				if balance > Zero::zero() {
+					Some((asset_id, balance))
+				} else {
+					None
+				}
+			})
+			.sorted_by(|a, b| b.1.cmp(&a.1))
+			.collect_vec()
+	}
+
+	pub fn funding_asset_to_ct_amount(
```

Comment:
1. Even though this won't be part of the consensus, it would be better to use `expect` instead of just `unwrap`. This way, the UI can reuse the error message inside the expect.
2. Let's remove the `dbg!` messages.
3. More conceptually: To avoid future mismatches, can we abstract the logic for computing the amount and reuse the same function in both the `do_contribute`/`do_bid` logic and here?

Reply:
in do_contribute/do_bid we have the CT amount and calculate the USD amount. In this API we do the reverse. So we cannot really abstract it